Feature/load cms

keith1111 opened the pull request:
changed content area display when on /cms , now displays processed data from database

keith1111 committed:
Add backend route articles

keith1111 committed:
Backend returns article JSON by ID from database

keith1111 committed:
Set up /cms frontend route to load data from backend

keith1111 committed:
Add topics tree to local storage

keith1111 committed:
Add summary view when on topic/subtopic

keith1111 committed:
Use marked for display articles in html

keith1111 committed:
Load topic tree to sidebar (demonstration mode)

keith1111 committed:
Linter app + minor styling fix ContentArea

nikrb commented on frontend/package.json:
```diff
@@ -9,6 +9,7 @@
     "prop-types": "^15.6.0",
     "react": "^16.2.0",
     "react-dom": "^16.2.0",
+    "react-router": "^4.2.0",
```
we don't need this me thinx

keith1111 replied:
me thinx too, worked without it,   reason is eslint.  if delete this line, npm run lint and we see error

jenovs replied:
You can import `Redirect` from ´react-router-dom` (`react-router` is its dependency) and get rid of this line.

nikrb commented on frontend/src/CMSContainer/CMSContainer.js:
```diff
@@ -1,13 +1,26 @@
+
+const CMSContainer = props => {
+  // takes data from database into local storage, if empty
+  checkLocalStorage();
```
I think we should populate local storage when app starts up.

keith1111 replied:
yes i'll move to another place

nikrb reviewed (changes requested):
are the all articles' fields store in localStorage?

nikrb commented on frontend/src/ContentArea/ContentArea.js:
```diff
@@ -1,43 +1,104 @@
-    this.state = {
-      innerHTML: undefined,
-    };
+  render() {
```
maybe refactor the viewy bits out to components?

keith1111 replied:
could not understand what you mean. type please which line needs refactor.

nikrb replied:
not "line", *function*. the render function is a bit long and hard to understand. Factoring out the view bits as components would help I think.

nikrb commented on frontend/src/ContentArea/ContentArea.js:
```diff
@@ -1,43 +1,104 @@
+    if (id) {
+      // exists id for that path
+      const index = JSON.parse(localStorage.getItem('articleIndex'));
+      const options = { __html: marked(articles[index[id]].content) };
```
getting picky now ... isn't this the html version of the article content? Why is the var called options?

keith1111 replied:
yes it is html code, but in view it displays raw html code with all tags on page.  To get around i use dangerouslySetInnerHtml in next line :  `view = <div dangerouslySetInnerHTML={options} />;`
dangerouslySetInnerHtml receives object with schema: `{__html : <html here>}`.    This object i call 'options'.   I think it is not critical so if you have another name for object, i listen. Anyway, we can call it MichaelSchumacher or moonWalker123 but it still would be object which we pass to dangerouslySetInerHtml.

nikrb replied:
it's fine I was only referring to the name. I didn't realise it was called options in the dox. Seems a bit odd. `options` will do fine

nikrb commented on frontend/src/CMSContainer/operate.js:
```diff
@@ -0,0 +1,69 @@
+}
+
+export default function checkLocalStorage() {
+  if (!localStorage.getItem('allArticles')) {
```
allArticles exists but pathTable doesn't. Should the pathTable key be tested for existence?

keith1111 replied:
my fault, i planned to 'chain' them so if  `allArticles` -> exists also `pathTable` and etc.  so check of `allArticles` would be enough.
but something went wrong. next commit will fix that.

keith1111 commented:
yes for details look local storage in browser tools

nikrb commented:
yes. I didn't think it would be a good idea to store the article content in localStorage ... but as I said in slack, I'm coming round to the idea from a mobile offline first perspective.
Of course something to think about not do now (using IndexedDb). A quick search suggests localStorage limited to 5 - 10MB, where as IndexedDb up to 500MB. Currently we only have about 151K.

keith1111 commented:
i saw too about 5 MB , that makes sad :(   so we keep in mind indexedDB

keith1111 committed:
Remove unnecessary dependencies frontend/package.json

keith1111 committed:
Move loading of articles to app start

keith1111 committed:
Crack ContentArea view into pieces (specific components)

keith1111 committed:
Linter

keith1111 committed:
Add detect of cms ready (asyncronous loading data)

keith1111 committed:
Add support articles without content and/or subtopic

keith1111 committed:
Add loading component CMSContainer

keith1111 committed:
Throw off ComponentWillMount

keith1111 committed:
Linter

nikrb commented on backend/routes/cms/getArticleJSON.js:
```diff
@@ -0,0 +1,24 @@
+  if (id && id.match(/^[0-9a-fA-F]{24}$/)) {
+    query = { _id: id };
+  }
+
```
I think we should respond with failure if the id is supplied not valid. Not a good idea to just return all users me things

keith1111 replied:
users?

keith1111 replied:
looked deeper if you mean all articles is bad,
request with 'id' is not in use by app. (frontend/src/actions.js line 108 is only place which requests from server)
that block of code will never run. can be safely deleted while refactoring stage.
I think that was for development if i start server and manually type request into address bar, to see that it returns json.

nikrb commented on frontend/src/App.js:
```diff
@@ -22,6 +22,7 @@ class App extends Component {
     articles: articlesMockData,
     isLoggedIn: true,
     user: this.guestUser,
+    cmsReady: !!localStorage.getItem('allArticles'),
```
This seems a bit over the top just as a check, is there not a better way? We load the data somewhere, can we not use that to set cmsReady?

keith1111 replied:
daefault false and then set to true when loaded?

nikrb replied:
well yeah. perhaps pass a function down (setCMSReady) to the component that loads the articles?

nikrb commented on frontend/src/actions.js:
```diff
@@ -17,4 +17,111 @@ const logout = () =>
+    .replace(/[^A-Za-z0-9-| ]/g, '')
+    .replace(/\s/g, '-');
+
+const DEFAULT_DISPLAY_SUBTOPIC = 'Uncategorized Articles';
```
do we need this? An article without a sub-topic should just be displayed as a leaf node under the topic, I don't want to see it listed under "uncategorised Articles"

keith1111 replied:
it is possible to make, first thought it will break existing order, i'll look but warn it can take time

nikrb replied:
not sure what 'existing order' you're referring to? Do you mean e.g.
testing|first article
testing|sub-topic|another article
testing|sub-topic2|yet another article

keith1111 replied:
yes i mean exactly that. every case has now the same pattern. but as i said earlier, i ll check what i can do.

nikrb commented on frontend/src/actions.js:
```diff
@@ -17,4 +17,111 @@ const logout = () =>
+        const tree = buildTopicTree(articles);
+        localStorage.setItem('tree', JSON.stringify(tree));
+        // this == <App />
+        this.setState({ cmsReady: true });
```
noooo ... only just spotted this little doozy. horrible, no. :grin: 
Return a promise and resolve true when load finishes - or false if fails.

keith1111 replied:
why horrible? is not allowed to use `this.setState` ?  promise cool but one note; data load runs once when i first visit the site. (local storage empty).  A hour or a week later, 2nd or 1000th time i visit site, local storage is not empty, data load does not run but to render cms i still need to check cms ready or not?

jenovs replied:
Functions should be as Independent as possible and `actions` has no business in state management.

nikrb commented on frontend/src/App.js:
```diff
@@ -37,6 +38,8 @@ class App extends Component {
         this.setState({ isLoggedIn: false, user: this.guestUser });
       });
+
+    actions.checkLocalStorage.call(this);
```
Don't do this please dude, see associated comment in actions where you call this.setState

jenovs commented on backend/routes/cms/getArticleJSON.js:
```diff
@@ -0,0 +1,24 @@
+      res.redirect('/api/v1/articles/');
+      return;
+    }
+
```
`await` without `catch`

jenovs commented on frontend/package.json:
```diff
@@ -9,6 +9,7 @@
     "prop-types": "^15.6.0",
     "react": "^16.2.0",
     "react-dom": "^16.2.0",
+    
```
Redundant line

jenovs commented on frontend/src/CMSContainer/CMSContainer.js:
```diff
@@ -1,13 +1,31 @@
+    </Wrapper>
+  );
+};
 
```
React way would be using ternary operator:
```js
{props.cmsReady ? <ContentArea path={path} /> : <Loading />}
```
Then you can get rid of `let content` and `if`

jenovs commented on frontend/src/ContentArea/ContentArea.js:
```diff
@@ -1,43 +1,39 @@
+  render() {
+    const { path } = this.props;
+    let view = 'none';
+    const id = JSON.parse(localStorage.getItem('pathTable'))[path] || '';
```
I'm not sure it's a good idea to put this in `render`.

nikrb replied:
agree.
We call getItem nine times - perhaps we just call them once at the top level (in App.js) and pass allArticles, pathTable, tree and articleIndex down using props?

jenovs commented on frontend/src/ContentArea/Root.js:
```diff
@@ -0,0 +1,31 @@
+      const obj = {
+        path: k,
+        name: root[k].titleDisplay,
+      };
```
Here I'd prefer `reduce` instead of `filter` + `map`. Also it could just return JSX instead of calling `map` again below.

jenovs commented on frontend/src/ContentArea/Topic.js:
```diff
@@ -0,0 +1,60 @@
+
+    return view;
+  }
+}
```
`reduce` instead of `filter` + `map`
Ternary operator instead of `if...else`

keith1111 committed:
Remove temporary

keith1111 committed:
Insert from feature/load-sidebar

keith1111 committed:
Change ContentArea view to display only articles

keith1111 committed:
Use function to set cms ready state

keith1111 committed:
Backend add error log

keith1111 committed:
CMSContainer ternary operator

keith1111 committed:
Pass allArticles data into props

keith1111 committed:
Pass articleIndex data into props

keith1111 committed:
Refactor unreachable

keith1111 committed:
Linter

nikrb commented on frontend/src/App.js:
```diff
@@ -37,6 +39,16 @@ class App extends Component {
+      if (loaded) {
+        this.setState({
+          articles: JSON.parse(localStorage.getItem('allArticles')),
+          articleIndex: JSON.parse(localStorage.getItem('articleIndex')),
```
should we store articleIndex in localstorage or create it after loading allArticles?

nikrb reviewed (approved):
looks good to me, conflicts need resolving though

nikrb commented on frontend/src/CMSContainer/CMSContainer.js:
```diff
@@ -1,13 +1,42 @@
+  ) : (
+    <Loading />
+  );
+
```
you know you can put this code directly in the render, no need for the content variable.

keith1111 committed:
Resolve merge conflicts

keith1111 committed:
Remove empty reference

keith1111 committed:
Fix prop types

keith1111 committed:
Fix path wait.png

keith1111 committed:
Fix set state top level

keith1111 committed:
Import loading picture

keith1111 committed:
Linter

nikrb committed:
review: merge develop and fix conflicts

nikrb committed:
fix: collapsible test

rename Sidebar.test.js to Collapsible.test.js
this also fixes the failing test in Sidebar

nikrb committed:
fix: linting

nikrb committed:
test: update Collapsible snapshots
